ipaserver/plugins/dogtag.py

```diff
@@ -1623,54 +1081,106 @@ def revoke_certificate(self, serial_number, revocation_reason=0):
+            )
+
         # Convert serial number to integral type from string to properly handle
-        # radix issues. Note: the int object constructor will properly handle large
-        # magnitude integral values by returning a Python long type when necessary.
+        # radix issues. Note: the int object constructor will properly handle
+        # large magnitude integral values by returning a Python long type
+        # when necessary.
         serial_number = int(serial_number, 0)
 
-        # Call CMS
-        http_status, _http_headers, http_body = \
-            self._sslget('/ca/agent/ca/doRevoke',
-                         self.env.ca_agent_port,
-                         op='revoke',
-                         revocationReason=revocation_reason,
-                         revokeAll='(certRecordId=%s)' % str(serial_number),
-                         totalRecordCount=1,
-                         xml='true')
+        path = 'agent/certs/{}/revoke'.format(serial_number)
+        data = '{{"reason":"{}"}}'.format(reasons[revocation_reason])
 
-        # Parse and handle errors
+        http_status, _http_headers, http_body = self._ssldo(
+            'POST', path,
+            headers={
+                'Content-Type': 'application/json',
+                'Accept': 'application/json',
+            },
+            body=data,
+            use_session=False,
+        )
         if http_status != 200:
             self.raise_certificate_operation_error('revoke_certificate',
                                                    detail=http_status)
 
-        parse_result = self.get_parse_result_xml(http_body, parse_revoke_cert_xml)
-        request_status = parse_result['request_status']
-        if request_status != CMS_STATUS_SUCCESS:
-            self.raise_certificate_operation_error('revoke_certificate',
-                                                   cms_request_status_to_string(request_status),
-                                                   parse_result.get('error_string'))
+        try:
+            response = json.loads(ipautil.decode_json(http_body))
+        except ValueError as e:
+            logger.debug("Response from CA was not valid JSON: %s", e)
+            raise errors.RemoteRetrieveError(
+                reason=_("Response from CA was not valid JSON")
+            )
+
+        request_status = response['operationResult']
+        if request_status != 'success':
+            self.raise_certificate_operation_error(
+                'revoke_certificate',
+                request_status,
+                response.get('errorMessage')
+            )
 
         # Return command result
         cmd_result = {}
 
-        cmd_result['revoked'] = parse_result.get('revoked') == 'yes'
+        # We can assume the revocation was successful because if it failed
+        # then REST will return a non-200 or operationalResult will not
+        # be 'success'.
+        cmd_result['revoked'] = True
```

Comment:
In the special case where revocation reason = 8 we see 'requestType': 'unrevocation' and I wonder if we should return cmd_result['revoked'] = False instead or even cmd_result['unrevoked'] = True.

Reply:
I checked the RFC more thoroughly on this state and found basically a state machine for generating a CRL. The notable bit is:

    6.3.3 CRL Processing

      (k)  If (cert_status is removeFromCRL), then set cert_status to
           UNREVOKED.

It is the CA that manages this state and it would set the status to UNREVOKED. It isn't something that IPA would do. 

In any case, at the point that we request removeFromCRL state it remains revoked until the CRL is processed. So I think True is correct, if not confusing.

Reply:
ok, thanks for the clarification